### ISSUE TITLE
More regular logging?

### ISSUE THREAD

**Issue by deiruch** [msoos/cryptominisat#430]
CryptoMiniSat produces a somewhat irregular output. Following is example output by version 5.0.1. While the output probably makes a lot of sense for the developer(s) of the solver, it is far less useful for me as a user. It appears that the various subsystems of the solver all produce output in a different format, intermixed with each other. The resulting irregularity makes it difficult to assess the progress (or lack thereof) of the solver. Other solvers like RISS or Lingeling produce a more regular output in a tabular fashion that I find easier to interpret.

```
c [simp] mem usage for occur   1229 MB
c Mem for watch alloc      : 356         MB (0.00      %)
c Mem for watch array      : 1           MB (0.00      %)
c --> Executing OCC strategy token: occ-backw-sub-str
c [sub] tri upI: 33131 subs w bin: 0 str w bin: 0 subs w tri: 0 str w tri: 0 tried: 1136781 str: 0 0-depth ass: 0 T: 0.04 T-out: N T-r: 98.35%
c [sub] rem cl: 0 tried: 7166284/1242526 (576.8%) T: 4.07 T-out: Y T-r: -0.00%
c [str] sub: 0 str: 0 tried: 3501339/1242526 (281.8)  T: 2.12 T-out: Y T-r: -0.00%
c [occ-substr] sub_str_with sub: 0 str: 0 0-depth ass: 0 T: 0.00 T-out: N T-r: 100.00%
c --> Executing OCC strategy token: occ-clean-implicit
c --> Executing OCC strategy token: occ-bve
c Empty resolvent elimed: 0 T: 0.02 T-out: N
c [occ-substr] sub_str_with sub: 0 str: 0 0-depth ass: 0 T: 0.00 T-out: N T-r: 100.00%
c [occ-bve] process impls_sub_lits  T: 0.00
c [occ-bve] iter v-elim 0
c  #try to eliminate: 23989
c  #var-elim: 0
c  #T-o: N
c  #T-r: 99.84%
c  #T: 0.64
c [occur] 2.33 is overhead
c [occur] link-in T: 2.33 cleanup T: 0.00
c --> Executing OCC strategy token: occ-bva
c [impl sub] bin: 0 tri: 0 (stamp: 0, cache: 0) T: 0.02 T-out: N w-visit: 48476
c [occ-bva] added: 0 simp: 0 2lit: Y T: 0.32 T-out: N T-r: 87.39%
c --> Executing OCC strategy token: occ-xor
c [impl str] lit bin: 0 (by stamp: 0) set-var: 0 T: 0.04 T-out: N T-r: 98.45% w-visit: 48476
c --> Executing strategy token: cache-clean
c [cache] cleaned. Updated:       0 K Cleaned:       0 K Freed:       0 K T: 0.05
c --> Executing strategy token: sub-str-cls-with-bin
c [clean] T: 0.4280 s
c [distill-with-bin-ext] stamp-based lit-rem: 0 inv-lit-rem: 0 stamp-cl-rem: 0
c [distill-with-bin-ext] bin-based lit-rem: 0 cl-sub: 0
c [distill-with-bin-ext] cache-based lit-rem: 0 cl-sub: 0
c [distill-with-bin-ext] T: 0.18 T-out: Y T-r: -0.00%
c [distill-with-bin-ext] stamp-based lit-rem: 0 inv-lit-rem: 0 stamp-cl-rem: 0
c [distill-with-bin-ext] bin-based lit-rem: 0 cl-sub: 0
c [distill-with-bin-ext] cache-based lit-rem: 0 cl-sub: 0
c [distill-with-bin-ext] T: 0.14 T-out: Y T-r: -0.22%
c [distill-with-bin-ext] stamp-based lit-rem: 0 inv-lit-rem: 0 stamp-cl-rem: 0
c [distill-with-bin-ext] bin-based lit-rem: 0 cl-sub: 0
c [distill-with-bin-ext] cache-based lit-rem: 0 cl-sub: 0
c [distill-with-bin-ext] T: 0.10 T-out: Y T-r: -0.03%
c [distill-with-bin-ext] stamp-based lit-rem: 0 inv-lit-rem: 0 stamp-cl-rem: 0
c [distill-with-bin-ext] bin-based lit-rem: 4 cl-sub: 0
c [distill-with-bin-ext] cache-based lit-rem: 3331 cl-sub: 0
c [distill-with-bin-ext] T: 0.12 T-out: Y T-r: -0.09%
c [distill] cache-based irred--  cl tried    37091 cl-sh     0 cl-rem    0 lit-rem      0 T: 0.28 T-out: Y
c [distill] cache-based   red--  cl tried      848 cl-sh   219 cl-rem    0 lit-rem   3335 T: 0.25 T-out: Y
c --> Executing strategy token: distill-cls
c [distill] longirred tried: 410/57919 cl-r:0 lit-r:0 T: 1.88 T-out: Y T-r: 100.08%
c [distill] tri+long useful: 0/410/57919 lits-rem: 0 0-depth-assigns: 0 T: 1.88 T-out: Y
c --> Executing strategy token: scc-vrepl
c [scc] new: 0 BP 1M T: 0.06
c --> Executing strategy token: check-cache-size
c --> Executing strategy token: renumber
c [cache] cleaned. Updated:       0 K Cleaned:       0 K Freed:       0 K T: 0.05
c [consolidate] T: 0.00
c clause size stats. size3: 47811 size4: 5607 size5: 798 larger: 3703
c 300-long burst search  learnt units:0 learnt bins: 0 T: 2.49
c   type VSIDS  rest  conf freevar  IrrL  IrrB l/longC  l/allC RedL0 RedL1 RedL2  RedB l/longC  l/allC
c   glue     1  4233 1358K   23989   57K   48K    3.32    2.71 1124K   46K 15898  123K   72.73   66.04
c [DBclean lev1] used recently: 1311 not used recently: 45271 moved w0: 476 T: 0.00
c [DBclean lev2] marked: 18350 ttl:0 locked_solver:0 T: 0.04
c   glue     1  4265 1365K   23989   57K   48K    3.32    2.71 1131K   41K 18350  123K   72.54   65.90
c [DBclean lev1] used recently: 2716 not used recently: 39073 moved w0: 370 T: 0.00
c   glue     1  4285 1372K   23989   57K   48K    3.32    2.71 1138K  2738   57K  123K   72.51   65.91
c   glue     1  4291 1379K   23989   57K   48K    3.32    2.71 1145K  2803   57K  123K   72.43   65.87
c   type VSIDS  rest  conf freevar  IrrL  IrrB l/longC  l/allC RedL0 RedL1 RedL2  RedB l/longC  l/allC
c [DBclean lev1] used recently: 2789 not used recently: 7 moved w0: 93 T: 0.00
c [DBclean lev2] marked: 17229 ttl:0 locked_solver:0 T: 0.05
c   glue     1  4316 1386K   23989   57K   48K    3.32    2.71 1152K   42K 17229  123K   72.34   65.82
c [DBclean lev1] used recently: 4002 not used recently: 37758 moved w0: 435 T: 0.01
```

For some inspiration, here's the output of other SAT solvers on the same problem instance. RISS first:
```
c WARNING: for repeatability, setting FPU to use double precision
c =================[ riss (core) 7.0.0                         cf1a66e ]===================================
c | Norbert Manthey. The use of the tool is limited to research only!                                     |
c | Based on Minisat 2.2 and Glucose 2.1  -- thanks!                                                      |
c | Contributors:                                                                                         |
c |      Kilian Gebhardt                                                                                  |
c |      Lucas Kahlert, Franziska Krüger, Aaron Stephan                                                   |
c ============================[ Problem Statistics ]=======================================================
c |                                                                                                       |
c |  Number of variables:              33676                                                              |
c |  Number of clauses:               156265                                                              |
c |  Number of total literals:        400573                                                              |
c |  Parse time:                        0.02 s                                                            |
c |                                                                                                       |
c ========================================[ MAGIC CONSTANTS ]==============================================
c | Constants are supposed to work well together :-)                                                      |
c | however, if you find better choices, please let us known...                                           |
c |-------------------------------------------------------------------------------------------------------|
c |                                |                                |                                     |
c | - Restarts:                    | - Reduce Clause DB:            | - Minimize Asserting:               |
c |   * LBD Queue    :     50      |   * First     :   4000         |    * size <  30                     |
c |   * Trail  Queue :   5000      |   * Inc       :    300         |    * lbd  <   6                     |
c |   * K            :   0.80      |   * Special   :   1000         |                                     |
c |   * R            :   1.40      |   * Protected :  (lbd)< 30     |                                     |
c |                                |                                |                                     |
c =========================================================================================================
c [STAT] CP3 3.14062 s-ppTime, 0 s-ipTime, 5.67278 s-ppwTime, 0 s-ipwTime, 569.207 MB, ok 0.03125 s-ohTime,
c [STAT] CP3(2) 119579 cls, 0 learnts, 28731 rem-cls, 0 rem-learnts,
c [STAT] UP 0 s, 0 units, 19316 cls, 11096 lits
c [STAT] SuSi(1) 0.046875 s, 4643 cls,  with 17347 lits, 4336 strengthed
c [STAT] SuSi(2) 2440077 subs-steps, 4951387 strenght-steps, 0 increases, 0.109375s strengthening
c [STAT] EE 0.125 s, 0 steps 865 ee-lits 2477 removedCls, 0 removedSubCls,
c [STAT] EE-gate 0.03125 s, 105591 steps, 0.0625 extractGateTime,
c [STAT] BVE(1) 0.25 s, 0.1875 s spent on subsimp, 39998 vars tested, 39997 anticipations, 7661 vars skipped, 652285 checks,
c [STAT] BVE(2) 48256 rem cls, with 145959 lits, 0 learnts rem, with 0 lits, 49256 new cls, with 194514 lits, 0 new learnts, with 0 lits,
c [STAT] BVE(3) 6035 vars eliminated, 0 units enqueued, 0 BC removed, with 0 lits, 0 blocked learnts removed, with 0 lits,
c [STAT] BVE(4) 37239 gateDefs, 5648 usedGates, 0 gateSeconds,
c [STAT] BVE(5) 2044 incElims, 894 keepElims, 3097 decElims, 1000 totalAdds,
c [STAT] UNHIDE 0.234375 s, 393 cls, 8664 totalLits, 7183 lits
c [STAT] UNHIDE(2) 0 prSecs, 0 prSteps, 0 L1units, 0 L2units, 0 L3units, 0 L4units, 0 L5units,
c [STAT] UNHIDE(3) 0 EE-checks, 0 EE-cands, 0 EEs
c [STAT] XOR 0.046875s, 0.046875 parse-s, 0 reason-s,
c [STAT] XOR(2) 832 xors, 0 subXors, 0 resXors, 0 clauses, 0 resClauses, 0 resTauts, 0 resStrength, 270890 findChecks,
c [STAT] XOR(3) 709 steps, 66056 empties, 0 listUnits, 62 useLists, 0 xorUnits, 222 eqs,
c [STAT] XOR(4) 122586 participating, 0.920087 ratio, 0 simDecisions, 0 xorProps, 0 clsProps,
c [STAT] BCE 0.078125 seconds, 1030195 steps, 88662 testLits, 0 remBCE, 0 BCMs, 0 BCMcands, 0 BCMlits,
c [STAT] CLE 3 remCLE, 0 cleUnits,
c [STAT] BCE-CLA 0 seconds, 0 steps, 0 testLits, 0 extClss, 0 extLits, 0 possibles,
c [STAT] FM 2.375 s, 0.015625 amo-s, 2.67188 fm-s, 0.015625 amt-s, 5917 amos, 0 amts, 0 sameUnits, 6 deducedUnits, 0 propUnits, 139247 cls, 1510502 steps, 6341198 searchSteps, 215021 addDups,
c [STAT] FM(2) 0 irregulars, 40559 pureAmoLits, 100006 diff, 709276 discards, 144489 addedCs, 112513 removedCs, 639 newAmos, 0 discardedNewAmos,
c [STAT] FM(3) 7105 newAmoBinCls, 15 newCls, 219 newAlos, 0 newAlks, 165 duplicates, 1 garbageCollecst,
c [STAT] FM(4) 0.015625 2PrdTime, 3541 2PrAmos, 18316 2PrLits, 0.015625 dedAloTime, 0 dedAlos,
c [STAT] FM(5) time: 0.078125 s,steps: 1322304 cards: 1670 failed: 771 extLits: 1677 redDegree: 26 probes: 8902 conflictProbes: 2408 savedPreCls: 91018 savedPostCls: 2784 units: 17804
c [STAT] DENSE 12428 gaps
c =========================================================================================================
c | solve clauses:       119579  solve variables:        21248                                            |
c =========================================================================================================
c ==================================[ Search Statistics (every  10000 conflicts) ]=========================
c |                                                                                                       |
c |          RESTARTS           |          ORIGINAL         |              LEARNT              | Progress |
c |       NB   Blocked  Avg Cfc |    Vars  Clauses Literals |   Red   Learnts    LBD2  Removed |          |
c =========================================================================================================
c |       16         0      625 |   33636   118771   336776 |     1     7908     1107     1882 |  0.188 % |
c |       58        53      344 |   33633   118751   336709 |     3     7126     2137    12650 |  0.202 % |
c |      114       112      263 |   33633   118751   336709 |     3    17126     2909    12650 |  0.202 % |
c |      178       150      224 |   33633   118751   336709 |     4    17996     3479    21780 |  0.202 % |
c |      245       202      204 |   33631   118742   336676 |     5    16256     3954    33514 |  0.212 % |
c |      307       233      195 |   33631   118742   336676 |     5    26256     4420    33514 |  0.212 % |
c |      386       285      181 |   33631   118742   336676 |     6    21952     4762    47818 |  0.212 % |
```

And here's Lingeling:
```
c Lingeling SAT Solver
c
c Version bbc 923038088b727dcc8bd03aa8dfb5013bd03bf8c4
c
c Copyright (C) 2010-2016 Armin Biere JKU Linz Austria.
c All rights reserved.
c
c released Thu Jul  7 10:21:35 CEST 2016
c compiled Mon Oct 23 14:00:59 DST 2017
c
c gcc (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
c -Wall -O3 -DNLGLOG -DNDEBUG -DNCHKSOL -DNLGLDRUPLIG -DNLGLYALSAT
c -DNLGLFILES -DNLGLDEMA
c Linux cn029 4.4.0-43-Microsoft x86_64
c
c reading input file test.dimacs
c no embedded options
c found 'p cnf 33676 159224' header
c read 33676 variables, 159224 clauses, 405840 literals in 0.00 seconds
c
c  seconds         irredundant            redundant clauses   glue   iterations"      MB    stability
c          variables clauses conflicts  large ternary binary     jlevel       jlevel'  agility     tlevel
c
c  S    0.0   31534   156767         0       0      0     0    0     0      0     0    8  23     0      0
c  P    1.0   27745   119888         0       0      0     0    0     0      0     0   11  23     0      0
c  S    2.0   23233   114559      1830    1186    462 14444    6   292     78   -28    9  10   999  15970
c  S    5.0   22983   113437     16842    4556   2822 10331    7   159   1054   -90    9  14   998   5876
c  S   10.0   22974    91952     40063    8931  15725 16621    7    89   3397   -24   10   8   977   5837
c  S   20.0   14933    66042     82859   18339  46212 15716    8    70   3151    -3   12  15   961   2799
c  S   30.0   14928    66041    137668   33388  48705 17363    8    77   5075     0   14  11   955   3196
c  S   40.0   14860    64388    182445   41434  50107 15901    9    78   5449     3   15  14   960   3955
```

(I personally prefer Lingeling's output format, because it includes "seconds", which RISS' one doesn't.)

**Comment by msoos** [comment]
Yeah, I am aware of this. The issue is the following: nobody really understands what all of this means. Not even the solver developers. Number of variables? Is that progress? No. We can remove variables, but it might just get slower. Number of clauses? Is that progress? Less is better? More? Somewhere in between? What size? Smaller only? Larger too? Average size? Distribution?

Bottom line is, if you asked me about progress, I can say: no idea. And there are no strong predictors for difficulty or a solution being close. I have played with this for over 7 years now and I can't tell, and neither can the other solver developers, except the very last 0.1s of the solve, we can see a meaningful difference. You probably don't want that, you want some progress indicator.

All progress indicators are absolutely wrong. If you want to, I can give you a fake percentage, that's what MiniSat used to do. People loved it. Meant nothing at all, ask the developers. People think the number of variables means something. It doesn't. Nothing AT ALL. I can give you a problem with <1000 variables that you will never be able to solve, and if you can, predictably, the NSA will be happy to pay over 100 million, in your pocket, right now. So... variables mean nothing. Clauses, same thing. And I could go on.

So, if you want a fake percentage,please create a shell script that comes up with a random percentage, and just cat that to your terminal. If you actually understand what these stats mean above in terms of progress, please tell that to the SAT community, I think your name will be forever etched to the science of SAT solving.

**Comment by msoos** [comment]
What I can help you with, is that in case you have a strong reason to know some stat because it's useful to you in a meaningful way, I can tell you how to parse that in a meaningful way.

Note that CryptoMinISat is the ONLY solver that actually puts ALMOST ALL of the above data in a STRUCTURED place, in a sqlite database. You can actually read it without grepping, it's all there. And A LOT MORE. Just check it out. It can dump over a gig of stats to an sqlite file. You can dig that for indicators. If you find something cool, please let me know, we can make a research paper out of it, win competitions, be employed at 400k/year, all the cool stuff :)

Mate

**Comment by deiruch** [comment]
Haha, what an awesome reply :)

Reporting the elapsed time at an easy-to-find location would already be helpful for my uses. And yes, sure, in general the reported numbers are hardly useful. But if I repeatedly solve closely related instances, the numbers start to become at least somewhat meaningful. (I noticed that in my current tests, Lingeling always solved my instances when the number of variables reached 7600. Percentages are useless, but #variables and #clauses seem to work in this case.)

Similarly, if I let a solver run for weeks and realize that no stats changed over the last day, I stop the solver with an (undeserved?) good conscience. If I can see some development in the numbers, I'll let it run a bit longer. But maybe that's plain stupid, for the reasons you just explained...

TL;DR - I like to look at Time/#Variables/#Clauses manually to get a rough impression of progress when solving very similar problem instances.

**Comment by msoos** [comment]
:)

No stats changed over a week... that's not correct. The number of clauses will change. As for time, you can find that, and over 170 more indicators (!!!). Please see here: https://github.com/msoos/cryptominisat/blob/master/cmsat_tablestructure.sql

If lingeling solved it at 7600 variables, that will not be the same for Riss, MiniSat, Glucose, CMS or any other solver. Also, CMS reports the number or clauses and variables very clearly (but not time... maybe I could add that as a TODO):

```
c   type VSIDS  rest  conf freevar  IrrL  IrrB l/longC  l/allC RedL0 RedL1 RedL2  RedB l/longC  l/allC
c   glue     1  4233 1358K   23989   57K   48K    3.32    2.71 1124K   46K 15898  123K   72.73   66.04
```

freevar is the number of free variables, irredB + irredundant binary clauses, redL + redundant long clauses (0..2 are different categories), etc. I am creating an issue right now about the time not being reported to the console.

Again, all of this, ad MUCH more (literally, over 170 more!) is in the SQLite database, reported more than once a second. I know we are visual creatures, so I have been building a visualization from the SQLite DB. It looks like this: https://www.msoos.org/sat_visualization/

You are more than welcome to work on it too :) If you are interested, I can hook you up. It's an almost-finished Docker container.

**Comment by deiruch** [comment]
Sorry for apparently being unclear: I never compare these numbers between different solvers. And I do realize that all the information is there already or can be obtained. I typically use SAT solvers to solve sub-problems and combine their console outputs on different network nodes with console output of the overarching algorithm. Adding databases, browsers and webservers (or containers) into the mix seems overly complicated. (By the way: The SAT visualizations are great to look at!)

It was just a suggestion to improve the **user experience** for simple people like me that enjoy looking at console output.
